File: 백준/Silver/17266. 어두운 굴다리/어두운 굴다리.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());
        int m = Integer.parseInt(br.readLine());

        int[] arr = new int[m];

        StringTokenizer st = new StringTokenizer(br.readLine());
        for (int i = 0; i < m; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }

        int left = 1;
        int right = n;
        int ans = 0;

        while (left <= right) {

            int mid = (left + right) / 2;

            int light = 0;
            for (int pos : arr) {

                if (light < pos - mid) {
                    break;
                }

                light = pos + mid;
            }

            if (light >= n) {
                ans = mid;
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }

        System.out.println(ans);
    }
}

File: 백준/Silver/17266. 어두운 굴다리/어두운 굴다리.py
n = int(input())
m = int(input())

arr = list(map(int, input().split()))

left, right = 1, n
ans = 0

while left <= right:

    mid = (left + right) // 2

    light = 0

    for pos in arr:
        if light < pos - mid:
            break
        light = pos + mid

    if light >= n:
        ans = mid
        right = mid - 1
    else:
        left = mid + 1

print(ans)
